### PR TITLE
FEAT: Test Runner - Kick Off Testsuite

### PR DESCRIPTION
**Pull request by bentvelj**
#### Which issue(s) does this change fix?
Adds functionality to kick off a testsuite by running a Fargate task with a set of environment variables, using a [runTask](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RunTask.html) query.

#### Why is this change necessary?
In order for customers to run their testsuites on Fargate, there needs to be functionality to make a runTask query to start a test-running task.

#### How does it address the issue?
This change contains the code to make a test-running runTask query

#### What side effects does this change have?
N/A

### NOTE: 

The general image used in the Test Runner Task Definition is not yet public, so testing this is properly is not yet possible. Once remaining features are added, and the general image is public, thorough end to end testing can be done. Given that we are merging with a feature branch, I think waiting for further implementation and end to end tests is reasonable.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    botocore.ClientError
+        If the describe_subnets call fails
+    """
+    ec2_client = boto_client_provider("ec2")
+    subnets = ec2_client.describe_subnets().get("Subnets")
+    return [subnet["SubnetId"] for subnet in subnets]
+
+
+def invoke_testsuite(
+    boto_client_provider: BotoProviderType,
```
Comment (hoffa) [inline samcli/lib/test_runner/invoke_testsuite.py:34]: How does the caller specify the region?

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    )
+
+    click.secho("Successfully kicked off testsuite!\n", fg="green")
+
+    # If await is specified, wait for the task to finish and the results.tar.gz to appear in the bucket
+    if do_await:
+        click.secho("Awaiting testsuite completion...\n", fg="yellow")
+
+        task_arn = response.get("tasks")[0].get("taskArn")
+        task_waiter = ecs_client.get_waiter("tasks_running")
```
Comment (hoffa) [inline samcli/lib/test_runner/invoke_testsuite.py:133]: These are only used in L140, any reason they're separate?

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    """
+
+    # If the customer specifies their own subnet, use it, otherwise query describe-subnets.
+
+    subnets = other_env_vars.get("subnets") or get_subnets(boto_client_provider)
+
+    container_env_vars = [
+        {"name": "BUCKET", "value": bucket},
+        {"name": "TEST_RUN_ID", "value": path_in_bucket},
+        {"name": "OPTIONS", "value": test_command_options},
```
Comment (hoffa) [inline samcli/lib/test_runner/invoke_testsuite.py:98]: `BUCKET` and `OPTIONS` are maybe too generic as env vars, not unreasonable someone to already be using something similar. Even prefixing everything with `TEST_` (or whatever) could be a better way to reduce risk.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    ]
+
+    for key, value in other_env_vars.items():
+        if key in ("BUCKET", "TEST_RUN_ID", "OPTIONS"):
```
Comment (hoffa) [inline samcli/lib/test_runner/invoke_testsuite.py:102]: To avoid the risk of typos (hence risk of forgetting something), we can define `container_env_vars` as a regular dict, and before passing to the ECS API, convert to whatever weird notation it requires:

```python
[{"name": k, "value": v} for k, v in container_env_vars.items()]
```

File: samcli/lib/test_runner/invoke_testsuite.py (hunk 1 above)
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:34]: Caller wouldn't specify the region directly, but would have the region and profile through context, and then construct a `boto_client_provider` to pass. Why would we need region passed directly?

File: samcli/lib/test_runner/invoke_testsuite.py (hunk 3 above)
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:98]: sounds good, I think even `TEST_RUNNER_BUCKET`

File: samcli/lib/test_runner/invoke_testsuite.py (hunk 2 above)
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:133]: Nope, I'll move the line down.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
+from typing import *
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:1]: Will do

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    -------
+    List[str]
+        A list of subnet ids associated with a customer account.
+
+    Raises
+    ------
+    botocore.ClientError
+        If the describe_subnets call fails
+    """
+    ec2_client = boto_client_provider("ec2")
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:28]: There is a `get_boto_client_provider_from_session_with_config`, I've used the `get_boto_client_provider_with_config` as per [Mehmet's recommendation](https://github.com/aws/aws-sam-cli/pull/4024#discussion_r914697665). Is that what you mean by "respect sessions"?

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    click.secho("Successfully kicked off testsuite!\n", fg="green")
+
+    # If await is specified, wait for the task to finish and the results.tar.gz to appear in the bucket
+    if do_await:
+        click.secho("Awaiting testsuite completion...\n", fg="yellow")
+
+        task_waiter = ecs_client.get_waiter("tasks_running")
+        task_arn = response.get("tasks")[0].get("taskArn")
+        task_waiter.wait(cluster=ecs_cluster, tasks=[task_arn])
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:134]: I do not see anything in the [documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs.html#ECS.Waiter.TasksRunning) about events. It seems that it just waits for a successful state.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    Raises
+    ------
+    botocore.ClientError
+        If get_subnets, run_task, or waiters fail
+    """
+
+    # If the customer specifies their own subnet, use it, otherwise query describe-subnets.
+
+    subnets = other_env_vars.get("subnets") or get_subnets(boto_client_provider)
```
Comment (hoffa) [inline samcli/lib/test_runner/invoke_testsuite.py:93]: Is `subnets` an environment variable? If yes, it should be consistent with the rest.

**Review by hoffa — APPROVED** [review]

File: samcli/lib/test_runner/invoke_testsuite.py (hunk 8 above)
Comment (acristin) [reply samcli/lib/test_runner/invoke_testsuite.py:93]: Why is the customer specifying subnets via an environment variable?

Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:93]: Yeah it is confusing. Originally the customer would specify a bunch of options (e.g. bucket & ecs cluster overrides) in this file, but those are now options for the `run` command and the input file is solely for environment varibales. Will change subnet overrides to an option, same as the others, not sure how I missed this.

**Review by jfuss — CHANGES_REQUESTED** [review]

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
+import click
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:1]: Ah fair enough, we can move that message to the caller.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                {
+                    "name": container_name,
+                    "environment": [{"name": key, "value": val} for key, val in container_env_vars.items()],
+                }
+            ]
+        },
+        taskDefinition=task_definition_arn,
+    )
+
+    click.secho("Successfully kicked off testsuite!\n", fg="green")
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:97]: Will do, noted for command PRs.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+    # If await is specified, wait for the task to finish and the results.tar.gz to appear in the bucket
+    if do_await:
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:100]: I think we're gonna scrap the await option altogether, maybe include a fast-succeed option as a stretch feature. But I think above 80% of our customers will want to wait for tests to run and see results by default

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+    for key in other_env_vars.keys():
+        if key in container_env_vars.keys():
+            raise ReservedEnvironmentVariableException(
+                f"{key} is a reserved environment variable, ensure that it is not present in your environment variables file."
+            )
+
+    container_env_vars.update(other_env_vars)
+
+    ecs_client = boto_client_provider("ecs")
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:80]: Sure, I'll change the param to an ecs client

File: tests/unit/lib/test_runner/test_invoke_testsuite.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            "ecs_cluster": "test-cluster-name",
+            "container_name": "test-container-name",
+            "task_definition_arn": "test-task-def-arn",
+            "other_env_vars": {"TEST_RUNNER_BUCKET": "oh-no"},
+            "test_command_options": "--max-fail=1",
+            "subnets": ["subnet-xxxxxxxxxxxxxxxxx"],
+            "do_await": False,
+        }
+
+        with self.assertRaises(ReservedEnvironmentVariableException):
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply tests/unit/lib/test_runner/test_invoke_testsuite.py:27]: Agreed, will do

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
+import click
+from typing import List
+from samcli.commands.exceptions import ReservedEnvironmentVariableException
+from samcli.lib.utils.boto_utils import BotoProviderType
+
+
+def invoke_testsuite(
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:7]: Is there a reason it should be in a class, if it's just a single function?

File: tests/unit/lib/test_runner/test_invoke_testsuite.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
+from botocore.exceptions import ClientError
+from unittest import TestCase
+from samcli.lib.test_runner.invoke_testsuite import invoke_testsuite
+from samcli.commands.exceptions import ReservedEnvironmentVariableException
+from samcli.lib.utils.boto_utils import get_boto_client_provider_with_config
+
+
+class Test_InvokeTestsuite(TestCase):
+    def test_specify_reserved_variable(self):
```
[the remark this answers is not in the document]
Comment (bentvelj) [reply tests/unit/lib/test_runner/test_invoke_testsuite.py:9]: Adding a `isidentifier` check in the customer specifies invalid variable names

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        reserved_vars = []
+        for key in other_env_vars.keys():
+            if key in container_env_vars.keys():
+                reserved_vars.append(key)
+            if not str.isidentifier(key):
+                raise ValueError(f"{key} is not a valid environment variable name.")
+
+        if len(reserved_vars) > 0:
```
Comment (acristin) [inline samcli/lib/test_runner/invoke_testsuite.py:87]: Nit: I think this is counter to PEP 8 guidance; should look more like `if reserved_vars`. This is something that might be best addressed as a followup pass to catch all code that is inconsistent with PEP 8, and then we can make a call on which ones to change.

File: tests/unit/lib/test_runner/test_invoke_testsuite.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+            "ecs_cluster": "test-cluster-name",
+            "container_name": "test-container-name",
+            "task_definition_arn": "test-task-def-arn",
+            "other_env_vars": {},
+            "test_command_options": "--max-fail=1",
+            "subnets": ["subnet-xxxxxxxxxxxxxxxxx"],
+        }
+
+    def test_invalid_var_names(self):
+
```
Comment (acristin) [inline tests/unit/lib/test_runner/test_invoke_testsuite.py:23]: Nit: Remove extra blank lines

**Review by acristin — APPROVED** [review]

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    """
+
+    container_env_vars = {
+        "TEST_RUNNER_BUCKET": bucket,
+        "TEST_COMMAND_OPTIONS": test_command_options,
+        "TEST_RUN_ID": path_in_bucket,
+    }
+
+    for key in other_env_vars.keys():
+        if key in container_env_vars.keys():
```
[the remark this answers is not in the document]
Comment (jfuss) [reply samcli/lib/test_runner/invoke_testsuite.py]: > I'm adding an `isidentifier` check to each anyway

Not sure I understand what this is. Could you summarize this for me? Mainly for my understanding on the trade offs you are making.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
+"""
+Kicks off a testsuite in a Fargate container
```
Comment (jfuss) [inline samcli/lib/test_runner/invoke_testsuite.py:2]: Are there going to be other types of `SuiteRunner`s? You may want to consider making an interface and extend it to be Fargate. This gives you the ability to make different `SuiteRunner`s and interchange them with a factory or something similar.

An example of this is our "provider" for getting different types of resources parsed out of templates: https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/providers/provider.py#L472.

I wouldn't require this now but something you may want to consider having in place, which could make this much more expandable in the future.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+"""
+from typing import List
+import logging
+from samcli.commands.exceptions import ReservedEnvironmentVariableException
+from samcli.lib.utils.colors import Colored
+
+LOG = logging.getLogger(__name__)
+
+
+class SuiteRunner:
```
Comment (jfuss) [inline samcli/lib/test_runner/invoke_testsuite.py:12]: If this `SuiteRunner` is specific to Fargate, I would indicate that in the class name here.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+                    {
+                        "name": container_name,
+                        "environment": [{"name": key, "value": val} for key, val in container_env_vars.items()],
+                    }
+                ]
+            },
+            taskDefinition=task_definition_arn,
+        )
+
+        LOG.info(Colored().yellow("=> Successfully kicked off testsuite, waiting for completion...\n"))
```
Comment (jfuss) [inline samcli/lib/test_runner/invoke_testsuite.py:113]: Nit: You can move `Colored()` into the `__init__` of the class over doing it here. This way the class (as a whole) has access to it and you don't have to create it multiple times. You can also make it optional parameter to allow easier mocking and allow a caller to change it (if required). 

I wouldn't block the PR on this but trying to give you some additional things you may want to consider.

File: samcli/lib/test_runner/invoke_testsuite.py [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+            "TEST_COMMAND_OPTIONS": test_command_options,
+            "TEST_RUN_ID": path_in_bucket,
+        }
+
+        reserved_vars = []
+        for key in other_env_vars.keys():
+            if key in container_env_vars.keys():
+                reserved_vars.append(key)
+            if not str.isidentifier(key):
+                raise ValueError(f"{key} is not a valid environment variable name.")
```
Comment (jfuss) [inline samcli/lib/test_runner/invoke_testsuite.py:89]: I don't recommend raising `ValueError` or any of the default python Error or Exceptions. Main reason is they become hard to catch and know the reason, as `ValueError` could be something thrown by the interpreter or your code. Which forces you to inspect the message in order to understand deeper. Instead, create an Error specific to this case. You can extend ValueError to give the caller the ability to catch all.

File: tests/unit/lib/test_runner/test_invoke_testsuite.py [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+            "test_command_options": "--max-fail=1",
+            "subnets": ["subnet-xxxxxxxxxxxxxxxxx"],
+        }
+
+    def test_invalid_var_names(self):
+        self.params["other_env_vars"] = {r"0variable_name": "value", r"\othervariable_name": "othervalue"}
+        with self.assertRaises(ValueError):
+            self.runner.invoke_testsuite(**self.params)
+
+    def test_specify_variable_bucket(self):
```
Comment (jfuss) [inline tests/unit/lib/test_runner/test_invoke_testsuite.py:27]: Nit: you can parameterize this test with the next two. Up to you, but calling it out incase you were unaware of test parameterization.

**Review by jfuss — CHANGES_REQUESTED** [review]

File: samcli/lib/test_runner/invoke_testsuite.py (hunk 19 above)
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:2]: Great idea, will do

File: samcli/lib/test_runner/invoke_testsuite.py (hunk 21 above)
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:113]: Ah, good point, I just did it how I saw it done elsewhere, but it makes sense with lots of LOG messages

File: samcli/lib/test_runner/invoke_testsuite.py (hunk 22 above)
Comment (bentvelj) [reply samcli/lib/test_runner/invoke_testsuite.py:89]: Will do, I'll make note of this for other aspects of code.

File: tests/unit/lib/test_runner/test_invoke_testsuite.py (hunk 23 above)
Comment (bentvelj) [reply tests/unit/lib/test_runner/test_invoke_testsuite.py:27]: I wasn't, but I am now! This is much better, very easy to expand.
